### DISCUSSION TITLE
Github desktop + LFS?

### DISCUSSION THREAD

**Discussion by FlorianHimsl** [community/community#44860] [Other Feature Feedback, Questions, & Ideas]
How do you get it to initialize using LFS? How do you tell it which files you want it to use with LFS.
If Github Desktop already comes with LFS then why does it not work conveniently already?

The documentary stuff on the website seems to only be relevant to other git apps not desktop.

**Comment by r2d2Proton** [comment]
It is a pain. I just went through this.

You need to install git lfs:
https://docs.github.com/en/repositories/working-with-files/managing-large-files/installing-git-large-file-storage

After installing from a command prompt:

You need to run: git lfs install
You can track files:
git lfs track *.zip
git lfs track Final_TT_ULT_Texture_Pack_pt1.zip
git lfs track Final_TT_ULT_Texture_Pack_pt2.zip

Add and commit files:
git add -A
git commit -m "..."

And push:
git lfs push --all origin

**Reply by FlorianHimsl** [reply]
But Github Desktop is supposed to already have lfs installed. It does however not seem to have a command prompt. https://docs.github.com/en/desktop/installing-and-configuring-github-desktop/configuring-and-customizing-github-desktop/about-git-large-file-storage-and-github-desktop

Do you use the program called Github Desktop?

**Reply by chris-c-thomas** [reply]
Check my comment below. GitHub Desktop installed Git LFS for you but it stills needs to be configured for your repository. You'll do that through the command line.

**Comment by chris-c-thomas** [comment]
To clarify, when you download GitHub Desktop (the GUI app) and run it for the first time it will download and install Git LFS on your system. You then need to also configure Git LFS to track files in the repository which likely hasn't occurred in your scenario yet.

This is the article for that: [Configure LFS from command line](https://docs.github.com/en/repositories/working-with-files/managing-large-files/configuring-git-large-file-storage)

**Reply by chris-c-thomas** [reply]
It's not much command line work. You just need to tell it which file to be part of LFS. GitHub uses Photoshop files as an example so running `git lfs track "*.psd"` will start tracking all the `.psd` files within your current directory as LFS.

They'll then be handled accordingly in GitHub Desktop, the app. So no need to run any of the `git add` or `git commit`, etc commands (the whole point of managing your repo through a GUI) and no need to run `git lfs install` either as the app did that for you.

**Reply by FlorianHimsl** [reply]
How do I open git lfs then? In Github desktop there is def no command line option.

**Reply by chris-c-thomas** [reply]
There's no git lfs to open. The command line terminal and GitHub Desktop will be two different applications. Are you on macOS or Windows?

You'll use your terminal to run these commands. Once that is set up and in place, you won't need to run those commands again unless you need to make a change, such as adding another directory to be tracked my lfs.

You'll use your GitHub Desktop app to manage everything else Git related from there on out.

**Comment by FlorianHimsl** [comment]
Windows? what do I use`? powershell?

**Comment by chris-c-thomas** [comment]
**Note:** I just want to make sure you actually have a need for Git's Large File Storage functionality. Will you have files that are over 100mb in size? If not, we can likely make your repo work without this. And I mean individual files over 100mb. The repository can definitely be over that without issue and the need for LFS.

## Part 1 - Install newest PowerShell (v7+), git, and git-lfs
If you can use the latest version of PowerShell 7.3.2 that would be great. Install instructions are located [here](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.3) for Windows. I'm on a new Mac so I hadn't installed PowerShell yet but will do that now to try and walk you through the process.

Okay, I have 7.3.2 installed now as well.

First let's check and make sure both `git` and `git-lfs` are installed with proper versions.

Run: `git --version` in the command line. It should return `git` and it's version installed.
Then run: `git lfs --version` and that should give the `git-lfs` version that's installed.

This is what I got:
![Screenshot 2023-02-05 at 4 23 27 PM](https://user-images.githubusercontent.com/22353987/216849327-f253cb28-bf7c-4da9-b58f-6ed1ec5fe452.png)
If for some reason one of those isn't installed then we have to get them installed. I'm 98% sure you're in the clear with GitHub Desktop being install already but if not let me know I'll walk you through that too.

## Part 2 - Decide which files and/or directories you want tracked by LFS.
Not all the files in your repository need to be tracked by `git lfs`. Normal code files for your project very often don't. It's typically binary assets like .pdfs, .xlxs, .psd, and other media or graphic heavy type stuff. It could really be anything though.

So let's say your the name of your repo is `project` and within project you have multiple files and folders. But only the big files live in the folder `large-files` and thus that's what we want to track for `git lfs`. 

From PowerShell you're going to want to make sure you're in the root of your repository `project` and we're just going to track the folder `large-files`. So the next command we run is `git lfs track 'large-files/**'` . This is now tracking all the files within that directory.

But oh wait. You forgot you had 3 Photoshop files in path `project/photoshop/file1.psd` and so on. In this case we'll also have to add those. So, we can go to the directory where the Photoshop files are and run `git lfs track *.psd` and that will track all the files within the photoshop directory with the `.psd` extension.

## Part 3 - Using GitHub Desktop App
Now, you and the GUI app which makes this a lot easier at times to understand what's going on by giving you menus and visualizations a long the way.

The good knews is that those files are not being tracked and managed by `git-lfs` so the GitHub Desktop app will handle the rest from here on out. Adding, committing, pushing, etc. Give something like this a shot and come back if you need more assistance. I'd love to keep help you through the process.

## Part 4 - Atlassian's Git tutorials and support pages
I'm just linking this because I found them to be the way I learned how to be proficient at git before just refining my process by working on teams where weird scenarios would arise.

[Git Tutorials and Training | Atlassian](https://www.atlassian.com/git/tutorials)

**Reply by FlorianHimsl** [reply]
In Powershell it says I have git lfs installed but not git, I didn't want to install git, I tried to but the setup had like 20 questions and I want to use github desktop anyways, do I really have to install git just to use the console?

Can I not just say something like 

git lfs track 'C:\dropbox\Dropbox\Ballfrog\Assets\Scenes\Main.unity'

?
Why does github desktop come with git lfs but not a way to initialize it? And yes my project has very large files so I need this.

**Reply by chris-c-thomas** [reply]
> In Powershell it says I have git lfs installed but not git, I didn't want to install git

You need to have git installed. That's a requirement of using GitHub Desktop. Even though you are using the GUI you need git installed behind the scenes.

GitHub desktop comes with the ability to utilize LFS but you must tell it what files to track in the beginning / in the future as more Unity files are added that you want.

> Why does github desktop come with git lfs but not a way to initialize it?

I don't have an answer for that one but if you choose to use the GUI you'll have to still utilize the command line at times for it. Other GUIs, such as GitKraken, appear to work in the same manner: https://www.gitkraken.com/learn/git/git-lfs

**Reply by FlorianHimsl** [reply]
I use Git Desktop right now for many projects and it works fine, and powershell says I don't have git installed, when I use any commands for git. Weird. Maybe I should try installing git again.

**Reply by kevincastejon** [reply]
It's true that "sometimes" you don't have to install git "standalone", Github Desktop is/were probably installing it under the hood..
I definitely remember Github Desktop asking me to initialize LFS on Unity repo creation, it looks like it does not anymore... Even when it did, if you said no, then they were no way to "change mind" and initialize it afterward (still talking bout the Github Desktop app).
Weird design indeed... maybe they just decide to wipe out this feature.
I never actually checked, at the time, if the large files (>100mo) were automatically detected and added to LFS but I guess they were.
It would be nice to see this feature back and better integrated into this very handy app. Having to mix command lines and GUI is painfull and should be easilly automatized.
